Step2

anjaemin opened the pull request.

anjaemin committed:
LoginController 추가
LoginAcceptanceTest 추가

anjaemin committed:
TODO 제거

anjaemin committed:
step2 구현
- QnaController 추가
- QnaService 로직 구현
- HomeController에 questions 추가
- html에서 action url 변경
- QnaServiceTest 추가
- QnaAcceptanceTest 추가

jojoldu commented on src/main/java/nextstep/web/LoginController.java:
```diff
@@ -0,0 +1,41 @@
+public class LoginController {
+    private UserService userService;
+
+    @Autowired
```
생성자만 있으면 됩니다 :)
Autowired가 없어도 스프링에서는 인잭션을 해줍니다 :)

jojoldu commented on src/main/java/nextstep/web/QnaController.java:
```diff
@@ -0,0 +1,65 @@
+
+    @GetMapping("/update/{id}/form")
+    public String updateForm(@PathVariable("id") long questionId, @LoginUser User loginUser, Model model) {
+        Question question = qnaService.findById(questionId)
```
qnaService 에서 레파지토리를 그대로 반환하는데요.
서비스에서 Optional과 orElseThrow를 처리하고, 컨트롤러에서는 Question  결과만 받는게 더 좋습니다 :)
지금은 서비스가 하는 역할이 전혀 없네요!

jojoldu commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -37,13 +38,19 @@ public Question create(User loginUser, Question question) {
     public Question update(User loginUser, long id, Question updatedQuestion) {
-        // TODO 수정 기능 구현
-        return null;
+        questionRepository.findById(id)
```
이 코드가 의미가 모호합니다 :)
찾고 나서 어떤 기능을 하고 있지 않습니다.
이 코드는 어떤 의미일까요?
그 의미를 다른 분들은 알수있을까요?

만약에 지금 업데이트 하고자 하는 Question 이 디비에 존재하는지 확인하기 위한 용도라면
사실상 필요없는 기능입니다.
JpaRepository는 **있으면 update, 없으면 save**합니다.

진짜 필요한 기능인지 한번 정리해보시는게 좋아보입니다 :)

jojoldu commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -55,12 +62,18 @@ public void deleteQuestion(User loginUser, long questionId) throws CannotDeleteE
+        Question question = questionRepository.findById(questionId)
+                .orElseThrow(EntityNotFoundException::new);
+        Answer answer = new Answer(loginUser, contents);
+        answer.toQuestion(question);
```
이 기능이 정상적으로 작동하나요?
이 기능이 정상작동할려면
Question 의 addAnswer 메소드를 호출해야만 합니다.
 answer.toQuestion(question); 만 하시면 Answer에서는 Question 을 가져올수 있지만, Question 에서 Answer을 가져올수 없습니다.
Question 의 메소드를 사용하시고  answer.toQuestion(question);는 삭제해주세요 :)

jojoldu commented on src/test/java/nextstep/web/QnaAcceptanceTest.java:
```diff
@@ -0,0 +1,103 @@
+        User loginUser = defaultUser();
+        ResponseEntity<String> response = basicAuthTemplate(loginUser).getForEntity("/questions/form", String.class);
+        softly.assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+        log.debug("body : {}", response.getBody());
```
이 과정에서는 꼭 **제거 하실 필욘 없습니다**만,
실제 서비스 회사로 가실때면 이렇게 테스트 코드에 로그를 찍으시면 안됩니다.

보통 한개의 도메인 (주문, 결제, 정산, 포인트, 쿠폰 등등 서비스의 하위 도메인들) 마다 테스트 메소드가 대략 1천개 정도 되는데요.
이 테스트들은 브랜치에 push될때마다 전체가 수행되어야만 합니다.
그러다보니 
1) 전체 테스트가 얼마나 빨리 수행되는지
2) 사람이 눈으로 확인하지 않고, 시스템이 검증할수 있는지

가 테스트 코드들에게 있어 중요합니다.

로그를 찍는다는 것은 **두가지 규칙을 모두 위배하게 됩니다**
속도도 느려지게 되고, **자꾸 눈으로 확인해야할 수도 있습니다**

로그로 찍기 보다는 assertThat으로 눈으로 보고 싶은 내용들 전체를 검증하는 연습을 해주세요 :)

jojoldu reviewed (changes requested):
@anjaemin 안녕하세요 재민님 :)
QNA 게시판 리뷰어인 이동욱입니다.

스프링에 대한 코드리뷰가 이번 과정부터 시작인지라, 순차적으로 리뷰를 남기고자 합니다.
첫 단계에서는 스프링과 JPA에 대한 전반적인 내용에 관련되서 리뷰를 남겼으니 참고해주세요 :)

주말에 커밋하시느라 고생하셨습니다!

anjaemin committed:
step2 피드백 반영
- 생성자에 있는 @Autowired 제거
- qnaService findById 수정, addAnswer 수정
- 답변 추가/삭제 기능 추가
- QnaAcceptanceTest에 있는 log출력부분 제거

anjaemin committed:
step2 피드백 반영
- question update/delete 시 로그인한 사용자가 작성한 글인지 확인하는 메소드 추가(checkQuestionOwner)

jojoldu commented on src/main/java/nextstep/web/QnaController.java:
```diff
@@ -0,0 +1,74 @@
+    @GetMapping("/show/{id}")
+    public String showQuestion(@PathVariable("id") long questionId, Model model) {
+        Question question = qnaService.findById(questionId);
+        List<Answer> answers = question.getAnswers();
```
**트랜잭션 범위를 벗어나서 자식을 가져오는건 위험합니다**
보통 자식테이블을 가져오는건 **레이지 로딩**방식인데요.
이 레이지 로딩이 가능하려면 **트랜잭션 범위에 있어야만 합니다**
근데 컨트롤러에선 트랜잭션이 없는데 어떻게 작동하느냐!
그건 바로 Open Session In View 옵션이 **켜져 있기 때문**입니다.
이 옵션이 ON 되어 있으면 트랜잭션 범위를 벗어나도 **자식을 가져올수 있습니다**
그리고 스프링부트에서는 기본 설정이 ON입니다.

근데 실제 서비스에서는 이 옵션을 OFF 합니다.
성능상 패널티가 심하기 때문인데요.
이 옵션을 OFF 하는 순간 이 코드는 **자식을 레이지 로딩으로 가져올수 없다**는 에러와 함께 작동하지 않게 됩니다.

무조건 트랜잭션 어노테이션이 지정되어있는 서비스에서 모든 값을 조립해서 반환하도록 구현해주세요 :)

jojoldu commented on src/test/java/nextstep/service/QnaServiceTest.java:
```diff
@@ -0,0 +1,114 @@
+    private QnaService qnaService;
+
+    @Test
+    public void create() {
```
현재 이 코드를 비롯해서 아래의 모든 테스트 코드들은 **무의미**합니다.
a라는 메소드를 호출하면 b를 반환하도록 mocking 하고,
b가 반환되었는지 검증하는데요.
이건 당연히 b가 나옵니다.
그리고 **실제 코드를 전혀 테스트 하지 않습니다**

테스트 코드를 작성하는 이유는 **내가 생각하지 못했던 이슈들이 있는지 검증**하기 위함입니다.

특히나 이렇게 DB를 사용하는 경우에는 **내가 작성한 코드가 잘 돌아가는지 꼭 테스트**해보셔야합니다.
Service의 코드가 오류가 나는게 분명함에도 **지금 이 테스트 코드에선 발견할수 없습니다**
Mocking으로 무조건 정해진 값을 반환하기 때문이죠.

MockitoJUnitRunner가 아닌

```java
@RunWith(SpringRunner.class)
@SpringBootTest
```

로 실제로 **데이터를 가지고** 테스트 해보세요 :)

jojoldu reviewed (changes requested):
@anjaemin 재민님 고생하셨습니다!

이번엔 좀 더 심화된 내용으로 코드리뷰를 남겼습니다.

Mocking 테스트를 하면 안되는경우, JPA에서의 트랜잭션 등등에 대해 문제가 될만한것들은 리뷰를 남겼으니 **꼭 수정해주세요**

anjaemin committed:
step2 피드백 반영
- QnaServiceTest 수정(실제 데이터로 테스트)
- 트랜잭션 관련 수정
- 사용자 확인 메소드 수정

anjaemin committed:
step2 피드백 반영
질문글 조회시 답변과 같이 조회하는 부분 수정(NamedEntityGraph 사용)

jojoldu commented on src/main/java/nextstep/domain/QuestionRepository.java:
```diff
@@ -1,7 +1,13 @@
 public interface QuestionRepository extends JpaRepository<Question, Long> {
     Iterable<Question> findByDeleted(boolean deleted);
+
+    @EntityGraph(value = "QuestionWithAnswer")
```
@EntityGraph 는 데이터를 Outer Join으로 가져옵니다.
이너조인이 필요할땐 JOIN FETCH를 쓰시면 됩니다 :)

JpaRepository에 잘 적용하셨습니다.
엔티티에 직접 쓰시기 보단 이렇게 사용하시는걸 추천드립니다.

jojoldu commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -33,18 +34,31 @@ public Question create(User loginUser, Question question) {
+        return answerRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("해당 답변을 찾을 수 없습니다."));
+    }
+
```
QnaService 전체 코드를 보고 남길 라인이 없어 여기에 남깁니다.
생성자 DI를 쓰지 않고 @Resource를 쓰셨는데요.
Spring 의 Bean 주입 역시 **생성자로 가능합니다**
생성자로 Repository등을 주입받도록 하시면 Autowired나 Resource가 필요없고, **테스트 짜기에도 훨씬 수월합니다**

스프링에서 **필드 인잭션은 안티 패턴**으로 보고 있기 때문에 사용을 금지하시길 바랍니다 :)

참고: http://www.mimul.com/pebble/default/2018/03/30/1522386129211.html

jojoldu commented on src/test/java/nextstep/service/QnaServiceTest.java:
```diff
@@ -1,114 +1,98 @@
-
-@RunWith(MockitoJUnitRunner.class)
+@RunWith(SpringRunner.class)
+@SpringBootTest
```
피드백 반영 💯 

여기서 한걸음만 더 나아가볼까요?

SpringBoot로 하는 JPA 테스트에선 항상 **생성된 모든 데이터를 지워야만 합니다**
지금 같은 경우 Question과 Answer, User가 되겠네요 :)

지우지 않게 되면 **전체 테스트를 돌릴때 테스트가 서로 침범하게 되어 실패하게 됩니디**

예를 들어 테스트 A에서 데이터1,2,3을 넣었고, 다음으로 실행되는 테스트 B에서는 데이터 4만 넣고 1개만 들었는지 검증한다면 데이터는 테스트 A에서 넣은것도 H2 디비에 남아있어 4개가 나오고 테스트 B는 실패합니다.

근데, 테스트 B만 돌릴때는 테스트 A가 실행되지 않았기 때문에 데이터4만 있는 상태이며 테스트가 통과합니다.

즉, 전체 테스트와 개별 테스트간에 서로 다른 결과가 나오게 됩니다.
그러니 @After 메소드를 만들어 태스트에 사용된 데이터들을 **테스트가 끝날떄마다 삭제될 수 있도록** 코드를 추가해보세요 :)

jojoldu reviewed (changes requested):
@anjaemin 재민님 고생하셨습니다 :)
피드백 반영 아주 잘해주셨습니다.
이번엔 좀 더 한단계 높은 내용을 코멘트로 남겼습니다.
점점 더 실전에 가까운 코드로 개선되가고 있으니 조금만 더 힘내주세요!

anjaemin committed:
step2 피드백 반영
- fetch join 사용으로 변경
- QnaService @Autowired 제거 생성자 주입으로 변경
- QnaServiceTest @Before(질문, 답변 생성), @After(@Before에서 생성한 질문, 답변 삭제) 추가
** QnaServiceTest에 @Before, @After를 추가하면서 기존에 있던 질문생성, 삭제하는 단위테스트를 주석처리해놨습니다.

jojoldu reviewed (approved):
@anjaemin 재민님 피드백 반영 잘해주셨습니다 :)
Step2는 마치겠습니다!
다음 스텝 진행해주세요!
